In [1]:
import pandas as pd
import numpy as np
from loading_weather_data_into_python_timeline import data_weather_loc, data_weather_latlon

In [2]:
pd.set_option('display.max_columns', None)

# ¿Cuántas visitas se realizaron en total?

## Exploración

In [3]:
df_visits = pd.read_csv(r'C:\Users\Chio\Documents\pruebas\test_houm\question-custom-takehome-project-cjvok56xt3-dataset\visits.csv')

In [4]:
df_visits.head()

scheduled_id  property_id                 begin_date  \
0        169548            1  2022-01-13T10:00:00-03:00   
1        184763            1  2022-01-26T18:00:00-03:00   
2        186092            1  2022-01-28T12:00:00-03:00   
3        182497            1  2022-01-23T11:00:00-03:00   
4        182396            1  2022-01-23T10:00:00-03:00   

                    end_date type_visit     status  
0  2022-01-13T12:00:00-03:00      Visit  Cancelled  
1  2022-01-26T20:00:00-03:00      Visit  Cancelled  
2  2022-01-28T14:00:00-03:00      Visit  Cancelled  
3  2022-01-23T13:00:00-03:00      Visit  Cancelled  
4  2022-01-23T12:00:00-03:00      Visit  Cancelled

In [5]:
df_visits['status'].unique()

array(['Cancelled', 'Done'], dtype=object)

In [6]:
df_visits['type_visit'].unique()

array(['Visit'], dtype=object)

In [7]:
df_visits[df_visits['status'] == 'Done'].head()

scheduled_id  property_id                 begin_date  \
9         181607            2  2022-01-20T13:00:00-03:00   
11        177048            3  2022-01-14T19:30:00-03:00   
18        164793            3  2022-01-10T13:00:00-03:00   
24        168308            5  2022-01-07T11:00:00-03:00   
29        171942            5  2022-01-09T16:00:00-03:00   

                     end_date type_visit status  
9   2022-01-20T15:00:00-03:00      Visit   Done  
11  2022-01-14T21:30:00-03:00      Visit   Done  
18  2022-01-10T15:00:00-03:00      Visit   Done  
24  2022-01-07T13:00:00-03:00      Visit   Done  
29  2022-01-09T18:00:00-03:00      Visit   Done

In [8]:
df_visits_done = df_visits[df_visits['status'] == 'Done']

## Respuesta

In [9]:
len(df_visits_done.index)

139

Se realizaron 139 visitas en total

# ¿Cuál es el promedio de propiedades por propietario?

## Exploración

In [10]:
df_users = pd.read_csv(r'C:\Users\Chio\Documents\pruebas\test_houm\question-custom-takehome-project-cjvok56xt3-dataset\users.csv')

In [11]:
df_users.head()

property_id  user_id first_name last_name               address
0           31        1  Josephine   Darakjy   4 B Blue Ridge Blvd
1           34        2        Art    Venere  8 W Cerritos Ave #54
2           48        3      Lenna  Paprocki           639 Main St
3           27        4    Donette    Foller          34 Center St
4           40        5     Simona   Morasca          3 Mcauley Dr

In [12]:
df_users['user_id'].value_counts(ascending=False)

1     1
2     1
59    1
58    1
57    1
     ..
26    1
25    1
24    1
23    1
80    1
Name: user_id, Length: 80, dtype: int64

### Respuesta

In [13]:
df_users['user_id'].value_counts().mean()

1.0

El promedio de propiedades por propietario es 1.

# ¿Cuál era la temperatura promedio de todas las visitas que realizó en la propiedad del propietario con ID 2?

## Exploración

In [14]:
df_user_id2 = df_users[df_users['user_id'] == 2]

In [15]:
df_user_id2.head()

property_id  user_id first_name last_name               address
1           34        2        Art    Venere  8 W Cerritos Ave #54

In [16]:
df_properties = pd.read_csv(r'C:\Users\Chio\Documents\pruebas\test_houm\question-custom-takehome-project-cjvok56xt3-dataset\properties.csv')

In [17]:
df_properties.head()

property_id    type_house  business_type  bedrooms  bathrooms  \
0            1  departamento         Rental         1          1   
1            2  departamento  Rental & Sale         1          1   
2            3  departamento         Rental         2          2   
3            4  departamento  Rental & Sale         3          2   
4            5  departamento  Rental & Sale         2          3   

   parking_lots  services  balcony   pool  latitude  longitude    localidad  \
0             1       3.0        0  False  4.870956 -74.058040        Chí­A   
1             2       NaN        1  False  4.623068 -74.074030  Teusaquillo   
2             2       NaN        0  False  4.723909 -74.042336      Usaquen   
3             1       0.0        0  False  4.721716 -74.057976         Suba   
4             2       0.0        1  False  4.688163 -74.043300      Usaquen   

           city                  region   country  
0  Cundinamarca  Región De Cundinamarca  Colombia  
1        Bogotá  Región De Cundinamarca  Colombia  
2        Bogotá  Región De Cundinamarca  Colombia  
3        Bogotá  Región De Cundinamarca  Colombia  
4        Bogotá  Región De Cundinamarca  Colombia

In [18]:
df_properties_id2 = df_properties[df_properties['property_id'] == 34]

In [19]:
df_properties_id2.head()

property_id    type_house business_type  bedrooms  bathrooms  \
33           34  departamento        Rental         1          1   

    parking_lots  services  balcony   pool  latitude  longitude    localidad  \
33             0       NaN        0  False  4.618363   -74.0735  Teusaquillo   

      city                  region   country  
33  Bogotá  Región De Cundinamarca  Colombia

In [20]:
df_visits[(df_visits['property_id'] == 34) & (df_visits['status'] == 'Done')].head()

scheduled_id  property_id                 begin_date  \
178        187080           34  2022-01-28T11:00:00-03:00   
179        186857           34  2022-01-28T11:30:00-03:00   

                      end_date type_visit status  
178  2022-01-29T01:00:00-03:00      Visit   Done  
179  2022-01-28T14:00:00-03:00      Visit   Done

In [21]:
count_visits = df_visits['property_id'][(df_visits['property_id'] == 34) & (df_visits['status'] == 'Done')].count()
count_visits

2

In [22]:
location = df_properties_id2['localidad'].values.tolist()
location

['Teusaquillo']

In [23]:
startdates = df_visits['begin_date'][(df_visits['property_id'] == 34) & (df_visits['status'] == 'Done')].values.tolist()
startdates

['2022-01-28T11:00:00-03:00', '2022-01-28T11:30:00-03:00']

#### Suposición
Se asume que los primeros 19 caracteres corresponden al horario oficial de Colombia y cada una de sus localidades

In [24]:
startdates = [w[:19] for w in startdates]
startdates

['2022-01-28T11:00:00', '2022-01-28T11:30:00']

In [25]:
enddates = df_visits['end_date'][(df_visits['property_id'] == 34) & (df_visits['status'] == 'Done')].values.tolist()
enddates

['2022-01-29T01:00:00-03:00', '2022-01-28T14:00:00-03:00']

In [26]:
enddates = [w[:19] for w in enddates]
enddates

['2022-01-29T01:00:00', '2022-01-28T14:00:00']

In [27]:
for visit in range(count_visits):
    weather = data_weather_loc(location[0], startdates[visit], enddates[visit])


 - Requesting weather : 
 - Running query URL:  https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline/Teusaquillo/2022-01-28T11:00:00/2022-01-29T01:00:00?&unitGroup=metric&contentType=csv&include=days&key=7CKN3MUH7QBN7HYWKNPY6S69K


 - Requesting weather : 
 - Running query URL:  https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline/Teusaquillo/2022-01-28T11:30:00/2022-01-28T14:00:00?&unitGroup=metric&contentType=csv&include=days&key=7CKN3MUH7QBN7HYWKNPY6S69K



In [28]:
df_weather_visit1 = pd.read_csv(r'C:\Users\Chio\Documents\pruebas\test_houm\prueba\pregunta_3\weather_Teusaquillo_2022-01-28T110000_2022-01-29T010000.csv', delimiter=';')
df_weather_visit2 = pd.read_csv(r'C:\Users\Chio\Documents\pruebas\test_houm\prueba\pregunta_3\weather_Teusaquillo_2022-01-28T113000_2022-01-28T140000.csv', delimiter=';')

In [29]:
df_weather_visit1.head()

name    datetime  tempmax  tempmin  temp  \
0  Teusaquillo, Bogotá, D.C., Colombia  2022-01-28     21.2      9.2  14.6   
1  Teusaquillo, Bogotá, D.C., Colombia  2022-01-29     20.3      6.0  13.4   

   feelslikemax  feelslikemin  feelslike  dew  humidity  precip  precipprob  \
0          21.2           8.3       14.5  9.7      75.3     4.0         100   
1          20.3           6.0       13.3  8.1      74.0     0.4         100   

   precipcover preciptype  snow  snowdepth  windgust  windspeed  winddir  \
0        45.83       rain     0          0      16.6       31.2    272.2   
1        12.50       rain     0          0      15.1       25.5    260.3   

   sealevelpressure  cloudcover  visibility  solarradiation  solarenergy  \
0            1025.5        54.3        10.1           272.7         23.8   
1            1025.1        49.3        11.7           318.2         27.8   

   uvindex  severerisk              sunrise               sunset  moonphase  \
0        9          30  2022-01-28T06:11:51  2022-01-28T18:06:47       0.91   
1       10          10  2022-01-29T06:11:56  2022-01-29T18:07:04       0.95   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
1  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
1  rain  80222099999,80222399999,SKBO,80234099999,SKVV

In [30]:
df_weather_visit2.head()

name    datetime  tempmax  tempmin  temp  \
0  Teusaquillo, Bogotá, D.C., Colombia  2022-01-28     21.2      9.2  14.6   

   feelslikemax  feelslikemin  feelslike  dew  humidity  precip  precipprob  \
0          21.2           8.3       14.5  9.7      75.3       4         100   

   precipcover preciptype  snow  snowdepth  windgust  windspeed  winddir  \
0        45.83       rain     0          0      16.6       31.2    272.2   

   sealevelpressure  cloudcover  visibility  solarradiation  solarenergy  \
0            1025.5        54.3        10.1           272.7         23.8   

   uvindex  severerisk              sunrise               sunset  moonphase  \
0        9          30  2022-01-28T06:11:51  2022-01-28T18:06:47       0.91   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV

In [31]:
df_weather_visit_id2 = pd.concat([df_weather_visit1, df_weather_visit2], ignore_index=True)

In [32]:
df_weather_visit_id2.head()

name    datetime  tempmax  tempmin  temp  \
0  Teusaquillo, Bogotá, D.C., Colombia  2022-01-28     21.2      9.2  14.6   
1  Teusaquillo, Bogotá, D.C., Colombia  2022-01-29     20.3      6.0  13.4   
2  Teusaquillo, Bogotá, D.C., Colombia  2022-01-28     21.2      9.2  14.6   

   feelslikemax  feelslikemin  feelslike  dew  humidity  precip  precipprob  \
0          21.2           8.3       14.5  9.7      75.3     4.0         100   
1          20.3           6.0       13.3  8.1      74.0     0.4         100   
2          21.2           8.3       14.5  9.7      75.3     4.0         100   

   precipcover preciptype  snow  snowdepth  windgust  windspeed  winddir  \
0        45.83       rain     0          0      16.6       31.2    272.2   
1        12.50       rain     0          0      15.1       25.5    260.3   
2        45.83       rain     0          0      16.6       31.2    272.2   

   sealevelpressure  cloudcover  visibility  solarradiation  solarenergy  \
0            1025.5        54.3        10.1           272.7         23.8   
1            1025.1        49.3        11.7           318.2         27.8   
2            1025.5        54.3        10.1           272.7         23.8   

   uvindex  severerisk              sunrise               sunset  moonphase  \
0        9          30  2022-01-28T06:11:51  2022-01-28T18:06:47       0.91   
1       10          10  2022-01-29T06:11:56  2022-01-29T18:07:04       0.95   
2        9          30  2022-01-28T06:11:51  2022-01-28T18:06:47       0.91   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
1  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   
2  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
1  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
2  rain  80222099999,80222399999,SKBO,80234099999,SKVV

### Suposición
Existen 3 campos que indican temperatura: tempmax, tempmin y temp se considera solo el campo temp, no las temperaturas maximas y minimas de la fecha consultada. También, como la descarga de los datos se definió en unidades métricas se puede asumir que la temperatura está indicada en °C.


In [33]:
df_weather_visit_id2['temp'].mean()

14.200000000000001

## Respuesta

La temperatura promedio de todas las visitas que se realizó en la propiedad del propietario con ID 2 fue de 14,2 °C.

# ¿Cuál es la temperatura promedio de las visitas para los días con lluvia?

## Exploración

In [34]:
df_properties['localidad'].unique()

array(['Chí\xadA', 'Teusaquillo', 'Usaquen', 'Suba', 'Chapinero', 'Funza',
       'Kennedy', 'Puente Aranda', 'Bosa', 'Fontibon', 'Mosquera',
       'Ciudad Bolivar', 'Soacha', 'Madrid', 'Santa Fe', 'Engativa',
       'Cajicá', 'Rafael Uribe Uribe'], dtype=object)

In [35]:
df_properties['localidad'] = df_properties['localidad'].replace('Chí\xadA','Chía', regex=True)

### Suposición
Se asumirá visita realizada, por lo tanto, marcada como Done

In [36]:
df_visits_done_properties = pd.merge(df_visits_done, df_properties, on=['property_id'], how='inner')
df_visits_done_properties['datetime'] = df_visits_done_properties['begin_date'].map(lambda x: x[:10])


In [37]:
df_visits_done_properties.head()

scheduled_id  property_id                 begin_date  \
0        181607            2  2022-01-20T13:00:00-03:00   
1        177048            3  2022-01-14T19:30:00-03:00   
2        164793            3  2022-01-10T13:00:00-03:00   
3        168308            5  2022-01-07T11:00:00-03:00   
4        171942            5  2022-01-09T16:00:00-03:00   

                    end_date type_visit status    type_house  business_type  \
0  2022-01-20T15:00:00-03:00      Visit   Done  departamento  Rental & Sale   
1  2022-01-14T21:30:00-03:00      Visit   Done  departamento         Rental   
2  2022-01-10T15:00:00-03:00      Visit   Done  departamento         Rental   
3  2022-01-07T13:00:00-03:00      Visit   Done  departamento  Rental & Sale   
4  2022-01-09T18:00:00-03:00      Visit   Done  departamento  Rental & Sale   

   bedrooms  bathrooms  parking_lots  services  balcony   pool  latitude  \
0         1          1             2       NaN        1  False  4.623068   
1         2          2             2       NaN        0  False  4.723909   
2         2          2             2       NaN        0  False  4.723909   
3         2          3             2       0.0        1  False  4.688163   
4         2          3             2       0.0        1  False  4.688163   

   longitude    localidad    city                  region   country  \
0 -74.074030  Teusaquillo  Bogotá  Región De Cundinamarca  Colombia   
1 -74.042336      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
2 -74.042336      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
3 -74.043300      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
4 -74.043300      Usaquen  Bogotá  Región De Cundinamarca  Colombia   

     datetime  
0  2022-01-20  
1  2022-01-14  
2  2022-01-10  
3  2022-01-07  
4  2022-01-09

In [38]:
count_visits = df_visits_done_properties['scheduled_id'].count()
count_visits

139

In [39]:
latitudes = df_visits_done_properties['latitude'].values.tolist()
# latitudes

In [40]:
latitudes = [str(x) for x in latitudes]
# latitudes

In [41]:
longitudes = df_visits_done_properties['longitude'].values.tolist()
longitudes = [str(x) for x in longitudes]
# longitudes

In [42]:
startdates = df_visits_done_properties['begin_date'].values.tolist()
startdates = [w[:19] for w in startdates]
# startdates

In [43]:
enddates = df_visits_done_properties['end_date'].values.tolist()
enddates = [w[:19] for w in enddates]
# enddates

### Observación
No encuentra Suba, Colombia como localidad (tampoco en consulta en la web se encuentra información bajo ese nombre) por lo que se trabaja con latitud y longitud.

#### Extracción de archivos del clima

In [44]:
for visit in range(count_visits):
    weather = data_weather_latlon(latitudes[visit], longitudes[visit], startdates[visit], enddates[visit], 4)

### Continuación exploración

In [45]:
import glob
import os

In [46]:
df_x = pd.read_csv(r'C:\Users\Chio\Documents\pruebas\test_houm\prueba\pregunta_4\weather_4.65809_-74.06184_2022-02-10T120000_2022-02-10T140000.csv', delimiter=';')

In [47]:
df_x.head()

name    datetime  tempmax  tempmin  temp  feelslikemax  \
0  4.65809,-74.06184  2022-02-10     20.2      8.3  13.6          20.2   

   feelslikemin  feelslike   dew  humidity  precip  precipprob  precipcover  \
0           8.3       13.5  10.3      82.7     7.1         100           50   

  preciptype  snow  snowdepth  windgust  windspeed  winddir  sealevelpressure  \
0       rain     0          0       7.2       23.9    145.2            1024.4   

   cloudcover  visibility  solarradiation  solarenergy  uvindex  severerisk  \
0        59.9         9.9           177.9         15.6        6          10   

               sunrise               sunset  moonphase  \
0  2022-02-10T06:11:48  2022-02-10T18:09:15       0.34   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV

In [48]:
path = r'C:\Users\Chio\Documents\pruebas\test_houm\prueba\pregunta_4'
all_files = glob.glob(os.path.join(path , "*.csv"))
df_weather_p4 = pd.concat((pd.read_csv(f, delimiter=';') for f in all_files), ignore_index=True)

In [49]:
df_weather_p4.head()

name    datetime  tempmax  tempmin  temp  feelslikemax  \
0  4.5879345,-74.127815  2022-02-26     18.3     11.2  14.6          18.3   
1  4.5879345,-74.127815  2022-02-27     18.2     12.0  14.0          18.2   
2   4.5907006,-74.21821  2022-01-31     20.8      4.2  12.9          20.8   
3   4.5907006,-74.21821  2022-02-15     20.0      9.0  14.2          20.0   
4    4.593542,-74.20581  2022-02-20     20.1     10.0  14.1          20.1   

   feelslikemin  feelslike   dew  humidity  precip  precipprob  precipcover  \
0          11.2       14.6  11.4      82.5     7.6         100        83.33   
1          12.0       14.0  12.0      88.5     9.2         100        66.67   
2           4.2       12.8   7.3      72.4     0.3         100        12.50   
3           8.5       14.1  10.5      81.0     4.5         100        66.67   
4          10.0       14.1  10.1      78.4     4.8         100        50.00   

  preciptype  snow  snowdepth  windgust  windspeed  winddir  sealevelpressure  \
0       rain   0.0        0.0       9.0       18.0    184.1            1026.1   
1       rain   0.0        0.0       7.9       20.8    220.3            1025.3   
2       rain   0.0        0.0      14.0       22.3    232.2            1024.9   
3       rain   0.0        0.0       8.6       25.5    198.6            1025.8   
4       rain   0.0        0.0      11.5       24.6    247.2            1024.8   

   cloudcover  visibility  solarradiation  solarenergy  uvindex  severerisk  \
0        77.5         9.6           196.3         17.0        8        10.0   
1        87.7         9.3           171.9         15.1        6        10.0   
2        63.2        10.6           304.7         26.5        9        10.0   
3        81.1         9.8            97.8          8.5        6        10.0   
4        70.1        10.6           256.3         22.2        7        10.0   

               sunrise               sunset  moonphase  \
0  2022-02-26T06:08:47  2022-02-26T18:10:02       0.89   
1  2022-02-27T06:08:29  2022-02-27T18:09:58       0.94   
2  2022-01-31T06:12:37  2022-01-31T18:08:11       1.00   
3  2022-02-15T06:11:42  2022-02-15T18:10:23       0.49   
4  2022-02-20T06:10:40  2022-02-20T18:10:29       0.58   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
1  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
2  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   
3  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
4  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
1  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
2  rain                   80222099999,80222399999,SKBO  
3  rain                   80222099999,80222399999,SKBO  
4  rain              80222099999,80222399999,SKBO,SKVV

In [50]:
df_weather_p4['conditions'].unique()

array(['Rain, Partially cloudy'], dtype=object)

#### Suposición
Se asume que si las condiciones mencionan 'Rain' se puede considerar como día con lluvia

In [51]:
df_rainy = df_weather_p4[df_weather_p4['conditions'].str.contains('Rain')]

In [52]:
df_rainy.head()

name    datetime  tempmax  tempmin  temp  feelslikemax  \
0  4.5879345,-74.127815  2022-02-26     18.3     11.2  14.6          18.3   
1  4.5879345,-74.127815  2022-02-27     18.2     12.0  14.0          18.2   
2   4.5907006,-74.21821  2022-01-31     20.8      4.2  12.9          20.8   
3   4.5907006,-74.21821  2022-02-15     20.0      9.0  14.2          20.0   
4    4.593542,-74.20581  2022-02-20     20.1     10.0  14.1          20.1   

   feelslikemin  feelslike   dew  humidity  precip  precipprob  precipcover  \
0          11.2       14.6  11.4      82.5     7.6         100        83.33   
1          12.0       14.0  12.0      88.5     9.2         100        66.67   
2           4.2       12.8   7.3      72.4     0.3         100        12.50   
3           8.5       14.1  10.5      81.0     4.5         100        66.67   
4          10.0       14.1  10.1      78.4     4.8         100        50.00   

  preciptype  snow  snowdepth  windgust  windspeed  winddir  sealevelpressure  \
0       rain   0.0        0.0       9.0       18.0    184.1            1026.1   
1       rain   0.0        0.0       7.9       20.8    220.3            1025.3   
2       rain   0.0        0.0      14.0       22.3    232.2            1024.9   
3       rain   0.0        0.0       8.6       25.5    198.6            1025.8   
4       rain   0.0        0.0      11.5       24.6    247.2            1024.8   

   cloudcover  visibility  solarradiation  solarenergy  uvindex  severerisk  \
0        77.5         9.6           196.3         17.0        8        10.0   
1        87.7         9.3           171.9         15.1        6        10.0   
2        63.2        10.6           304.7         26.5        9        10.0   
3        81.1         9.8            97.8          8.5        6        10.0   
4        70.1        10.6           256.3         22.2        7        10.0   

               sunrise               sunset  moonphase  \
0  2022-02-26T06:08:47  2022-02-26T18:10:02       0.89   
1  2022-02-27T06:08:29  2022-02-27T18:09:58       0.94   
2  2022-01-31T06:12:37  2022-01-31T18:08:11       1.00   
3  2022-02-15T06:11:42  2022-02-15T18:10:23       0.49   
4  2022-02-20T06:10:40  2022-02-20T18:10:29       0.58   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
1  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
2  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   
3  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
4  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
1  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
2  rain                   80222099999,80222399999,SKBO  
3  rain                   80222099999,80222399999,SKBO  
4  rain              80222099999,80222399999,SKBO,SKVV

In [53]:
df_visits_done_properties['name'] = df_visits_done_properties['latitude'].astype(str) + ',' + df_visits_done_properties['longitude'].astype(str)


In [54]:
df_visits_done_properties.head()

scheduled_id  property_id                 begin_date  \
0        181607            2  2022-01-20T13:00:00-03:00   
1        177048            3  2022-01-14T19:30:00-03:00   
2        164793            3  2022-01-10T13:00:00-03:00   
3        168308            5  2022-01-07T11:00:00-03:00   
4        171942            5  2022-01-09T16:00:00-03:00   

                    end_date type_visit status    type_house  business_type  \
0  2022-01-20T15:00:00-03:00      Visit   Done  departamento  Rental & Sale   
1  2022-01-14T21:30:00-03:00      Visit   Done  departamento         Rental   
2  2022-01-10T15:00:00-03:00      Visit   Done  departamento         Rental   
3  2022-01-07T13:00:00-03:00      Visit   Done  departamento  Rental & Sale   
4  2022-01-09T18:00:00-03:00      Visit   Done  departamento  Rental & Sale   

   bedrooms  bathrooms  parking_lots  services  balcony   pool  latitude  \
0         1          1             2       NaN        1  False  4.623068   
1         2          2             2       NaN        0  False  4.723909   
2         2          2             2       NaN        0  False  4.723909   
3         2          3             2       0.0        1  False  4.688163   
4         2          3             2       0.0        1  False  4.688163   

   longitude    localidad    city                  region   country  \
0 -74.074030  Teusaquillo  Bogotá  Región De Cundinamarca  Colombia   
1 -74.042336      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
2 -74.042336      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
3 -74.043300      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
4 -74.043300      Usaquen  Bogotá  Región De Cundinamarca  Colombia   

     datetime                  name  
0  2022-01-20   4.6230683,-74.07403  
1  2022-01-14  4.7239094,-74.042336  
2  2022-01-10  4.7239094,-74.042336  
3  2022-01-07    4.6881633,-74.0433  
4  2022-01-09    4.6881633,-74.0433

#### Suposición
Se asume que datetime de los datos del clima es igual que begin_date de la tabla de visitas

In [55]:
df_done_prop_rainy = pd.merge(df_visits_done_properties, df_rainy, on=['name', 'datetime'], how='inner')

In [56]:
df_done_prop_rainy.head()

scheduled_id  property_id                 begin_date  \
0        181607            2  2022-01-20T13:00:00-03:00   
1        177048            3  2022-01-14T19:30:00-03:00   
2        164793            3  2022-01-10T13:00:00-03:00   
3        168308            5  2022-01-07T11:00:00-03:00   
4        171942            5  2022-01-09T16:00:00-03:00   

                    end_date type_visit status    type_house  business_type  \
0  2022-01-20T15:00:00-03:00      Visit   Done  departamento  Rental & Sale   
1  2022-01-14T21:30:00-03:00      Visit   Done  departamento         Rental   
2  2022-01-10T15:00:00-03:00      Visit   Done  departamento         Rental   
3  2022-01-07T13:00:00-03:00      Visit   Done  departamento  Rental & Sale   
4  2022-01-09T18:00:00-03:00      Visit   Done  departamento  Rental & Sale   

   bedrooms  bathrooms  parking_lots  services  balcony   pool  latitude  \
0         1          1             2       NaN        1  False  4.623068   
1         2          2             2       NaN        0  False  4.723909   
2         2          2             2       NaN        0  False  4.723909   
3         2          3             2       0.0        1  False  4.688163   
4         2          3             2       0.0        1  False  4.688163   

   longitude    localidad    city                  region   country  \
0 -74.074030  Teusaquillo  Bogotá  Región De Cundinamarca  Colombia   
1 -74.042336      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
2 -74.042336      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
3 -74.043300      Usaquen  Bogotá  Región De Cundinamarca  Colombia   
4 -74.043300      Usaquen  Bogotá  Región De Cundinamarca  Colombia   

     datetime                  name  tempmax  tempmin  temp  feelslikemax  \
0  2022-01-20   4.6230683,-74.07403     20.5     10.0  14.5          20.5   
1  2022-01-14  4.7239094,-74.042336     18.2      9.0  13.4          18.2   
2  2022-01-10  4.7239094,-74.042336     18.5      7.9  13.3          18.5   
3  2022-01-07    4.6881633,-74.0433     21.3      7.0  13.0          21.3   
4  2022-01-09    4.6881633,-74.0433     20.1     11.0  14.9          20.1   

   feelslikemin  feelslike   dew  humidity  precip  precipprob  precipcover  \
0           9.1       14.5   6.8      62.6     0.6         100        20.83   
1           8.5       13.4  10.4      83.4     3.8         100        41.67   
2           6.8       13.2  10.0      82.0     2.2         100        33.33   
3           5.6       12.8   5.0      63.3     0.8         100        29.17   
4          11.0       14.9  10.4      76.8     1.2         100        50.00   

  preciptype  snow  snowdepth  windgust  windspeed  winddir  sealevelpressure  \
0       rain   0.0        0.0      18.7       22.9    102.5            1025.7   
1       rain   0.0        0.0       6.5       17.6    219.4            1026.9   
2       rain   0.0        0.0      13.0       18.2    229.0            1025.8   
3       rain   NaN        NaN       NaN       22.0    155.1            1026.0   
4       rain   NaN        NaN       NaN       23.7    244.3            1026.6   

   cloudcover  visibility  solarradiation  solarenergy  uvindex  severerisk  \
0        65.7        11.3           289.4         25.1        9        10.0   
1        83.6        10.5           168.3         14.9        5        10.0   
2        72.2        10.7           241.0         21.0        7        10.0   
3        66.8        10.8           234.1         20.2        9         NaN   
4        58.1        10.2           222.4         19.0        8         NaN   

               sunrise               sunset  moonphase  \
0  2022-01-20T06:10:37  2022-01-20T18:04:14       0.53   
1  2022-01-14T06:09:06  2022-01-14T18:01:34       0.45   
2  2022-01-10T06:07:49  2022-01-10T17:59:48       0.28   
3  2022-01-07T06:06:40  2022-01-07T17:58:28       0.14   
4  2022-01-09T06:07:24  2022-01-09T17:59:25       0.24   

               conditions               

## Resultado

In [57]:
df_done_prop_rainy['temp'].mean()

14.385981308411182

La temperatura promedio de las visitas para los días con lluvia fue de 14,39 °C.

# ¿Cuál es la temperatura promedio para las visitas realizadas en la localidad de Suba?

## Exploración
### Observación:
Suba debe buscarse a través de latitud y longitud

In [58]:
df_prop_suba = df_properties[df_properties['localidad'] == 'Suba']

In [59]:
df_prop_suba.head()

property_id    type_house  business_type  bedrooms  bathrooms  \
3             4  departamento  Rental & Sale         3          2   
8             9  departamento         Rental         3          1   
9            10  departamento         Rental         1          2   
18           19  departamento         Rental         3          2   
27           28  departamento         Rental         1          1   

    parking_lots  services  balcony   pool  latitude  longitude localidad  \
3              1       0.0        0  False  4.721716 -74.057976      Suba   
8              0       3.0        0  False  4.733718 -74.107970      Suba   
9              1       NaN        0  False  4.723203 -74.053140      Suba   
18             1       NaN        0  False  4.767047 -74.048990      Suba   
27             0       3.0        0  False  4.753546 -74.093185      Suba   

      city                  region   country  
3   Bogotá  Región De Cundinamarca  Colombia  
8   Bogotá  Región De Cundinamarca  Colombia  
9   Bogotá  Región De Cundinamarca  Colombia  
18  Bogotá  Región De Cundinamarca  Colombia  
27  Bogotá  Región De Cundinamarca  Colombia

In [60]:
df_visits_done.head()

scheduled_id  property_id                 begin_date  \
9         181607            2  2022-01-20T13:00:00-03:00   
11        177048            3  2022-01-14T19:30:00-03:00   
18        164793            3  2022-01-10T13:00:00-03:00   
24        168308            5  2022-01-07T11:00:00-03:00   
29        171942            5  2022-01-09T16:00:00-03:00   

                     end_date type_visit status  
9   2022-01-20T15:00:00-03:00      Visit   Done  
11  2022-01-14T21:30:00-03:00      Visit   Done  
18  2022-01-10T15:00:00-03:00      Visit   Done  
24  2022-01-07T13:00:00-03:00      Visit   Done  
29  2022-01-09T18:00:00-03:00      Visit   Done

In [61]:
df_suba_done = pd.merge(df_prop_suba, df_visits_done, on=['property_id'], how='inner')

In [62]:
df_suba_done['name'] = df_suba_done['latitude'].astype(str) + ',' + df_suba_done['longitude'].astype(str)

In [63]:
df_suba_done.head()

property_id    type_house business_type  bedrooms  bathrooms  parking_lots  \
0           10  departamento        Rental         1          2             1   
1           19  departamento        Rental         3          2             1   
2           19  departamento        Rental         3          2             1   
3           28  departamento        Rental         1          1             0   
4           28  departamento        Rental         1          1             0   

   services  balcony   pool  latitude  longitude localidad    city  \
0       NaN        0  False  4.723203 -74.053140      Suba  Bogotá   
1       NaN        0  False  4.767047 -74.048990      Suba  Bogotá   
2       NaN        0  False  4.767047 -74.048990      Suba  Bogotá   
3       3.0        0  False  4.753546 -74.093185      Suba  Bogotá   
4       3.0        0  False  4.753546 -74.093185      Suba  Bogotá   

                   region   country  scheduled_id                 begin_date  \
0  Región De Cundinamarca  Colombia        167764  2022-01-04T13:00:00-03:00   
1  Región De Cundinamarca  Colombia        176814  2022-01-14T15:00:00-03:00   
2  Región De Cundinamarca  Colombia        166495  2022-01-03T15:30:00-03:00   
3  Región De Cundinamarca  Colombia        178497  2022-01-17T15:30:00-03:00   
4  Región De Cundinamarca  Colombia        177764  2022-01-17T17:00:00-03:00   

                    end_date type_visit status                 name  
0  2022-01-04T15:00:00-03:00      Visit   Done  4.7232027,-74.05314  
1  2022-01-14T17:00:00-03:00      Visit   Done  4.7670474,-74.04899  
2  2022-01-03T17:30:00-03:00      Visit   Done  4.7670474,-74.04899  
3  2022-01-17T18:00:00-03:00      Visit   Done  4.753546,-74.093185  
4  2022-01-17T19:00:00-03:00      Visit   Done  4.753546,-74.093185

In [64]:
list_propid_suba_done = df_suba_done['property_id']

In [65]:
list_propid_suba_done = list_propid_suba_done.tolist()

In [66]:
list_propid_suba = df_suba_done['property_id']
list_propid_suba = list_propid_suba.tolist()

In [67]:
df_suba_done['property_id'].count()

35

In [68]:
df_visits[~(df_visits['property_id'].isin(list_propid_suba)) & (df_visits['status'] == 'Done')].count()

scheduled_id    104
property_id     104
begin_date      104
end_date        104
type_visit      104
status          104
dtype: int64

In [69]:
df_suba_done['datetime'] = df_suba_done['begin_date'].map(lambda x: x[:10])

In [70]:
df_suba_done.head()

property_id    type_house business_type  bedrooms  bathrooms  parking_lots  \
0           10  departamento        Rental         1          2             1   
1           19  departamento        Rental         3          2             1   
2           19  departamento        Rental         3          2             1   
3           28  departamento        Rental         1          1             0   
4           28  departamento        Rental         1          1             0   

   services  balcony   pool  latitude  longitude localidad    city  \
0       NaN        0  False  4.723203 -74.053140      Suba  Bogotá   
1       NaN        0  False  4.767047 -74.048990      Suba  Bogotá   
2       NaN        0  False  4.767047 -74.048990      Suba  Bogotá   
3       3.0        0  False  4.753546 -74.093185      Suba  Bogotá   
4       3.0        0  False  4.753546 -74.093185      Suba  Bogotá   

                   region   country  scheduled_id                 begin_date  \
0  Región De Cundinamarca  Colombia        167764  2022-01-04T13:00:00-03:00   
1  Región De Cundinamarca  Colombia        176814  2022-01-14T15:00:00-03:00   
2  Región De Cundinamarca  Colombia        166495  2022-01-03T15:30:00-03:00   
3  Región De Cundinamarca  Colombia        178497  2022-01-17T15:30:00-03:00   
4  Región De Cundinamarca  Colombia        177764  2022-01-17T17:00:00-03:00   

                    end_date type_visit status                 name  \
0  2022-01-04T15:00:00-03:00      Visit   Done  4.7232027,-74.05314   
1  2022-01-14T17:00:00-03:00      Visit   Done  4.7670474,-74.04899   
2  2022-01-03T17:30:00-03:00      Visit   Done  4.7670474,-74.04899   
3  2022-01-17T18:00:00-03:00      Visit   Done  4.753546,-74.093185   
4  2022-01-17T19:00:00-03:00      Visit   Done  4.753546,-74.093185   

     datetime  
0  2022-01-04  
1  2022-01-14  
2  2022-01-03  
3  2022-01-17  
4  2022-01-17

In [71]:
startdates = df_suba_done['begin_date'].values.tolist()
startdates = [w[:19] for w in startdates]
# startdates

In [72]:
count_visits = df_suba_done['scheduled_id'].count()
count_visits

35

In [73]:
latitudes = df_suba_done['latitude'].values.tolist()
latitudes = [str(x) for x in latitudes]

In [74]:
longitudes = df_suba_done['longitude'].values.tolist()
longitudes = [str(x) for x in longitudes]

In [75]:
enddates = df_suba_done['end_date'].values.tolist()
enddates = [w[:19] for w in enddates]

In [76]:
for visit in range(count_visits):
    weather = data_weather_latlon(latitudes[visit], longitudes[visit], startdates[visit], enddates[visit], 5)

In [77]:
path = r'C:\Users\Chio\Documents\pruebas\test_houm\prueba\pregunta_5'
all_files = glob.glob(os.path.join(path , "*.csv"))
df_weather_p5 = pd.concat((pd.read_csv(f, delimiter=';') for f in all_files), ignore_index=True)

In [78]:
df_weather_p5.head()

name    datetime  tempmax  tempmin  temp  feelslikemax  \
0   4.689937,-74.05933  2022-02-12     21.0     11.0  15.3          21.0   
1  4.7085752,-74.05685  2022-02-15     20.1      9.0  14.2          20.1   
2  4.7085752,-74.05685  2022-02-16     20.1     11.0  15.1          20.1   
3  4.7085752,-74.05685  2022-02-17     19.9     11.9  15.1          19.9   
4  4.7085752,-74.05685  2022-02-20     20.1     10.0  14.1          20.1   

   feelslikemin  feelslike   dew  humidity  precip  precipprob  precipcover  \
0          11.0       15.3  10.6      76.3     5.1         100        54.17   
1           8.5       14.2  10.6      81.1     4.5         100        66.67   
2          11.0       15.1  10.0      73.9     4.0         100        37.50   
3          11.9       15.1  10.8      77.6     1.7         100        29.17   
4           9.1       14.1  10.1      78.5     4.8         100        50.00   

  preciptype  snow  snowdepth  windgust  windspeed  winddir  sealevelpressure  \
0       rain   0.0        0.0       9.0       24.3    197.9            1024.2   
1       rain   0.0        0.0       8.6       25.4    201.8            1025.7   
2       rain   0.0        0.0      14.4       22.2    216.0            1026.0   
3       rain   0.0        0.0      10.4       16.5    160.4            1024.9   
4       rain   0.0        0.0      11.5       23.2    247.1            1024.9   

   cloudcover  visibility  solarradiation  solarenergy  uvindex  severerisk  \
0        71.1        10.0           209.4         18.4        7        30.0   
1        80.4         9.6            97.8          8.5        6        10.0   
2        67.5        10.1           251.3         21.9        7        10.0   
3        74.9        10.2           244.3         21.4        8        10.0   
4        70.3        10.4           256.3         22.2        7        10.0   

               sunrise               sunset  moonphase  \
0  2022-02-12T06:11:36  2022-02-12T18:09:26       0.42   
1  2022-02-15T06:11:10  2022-02-15T18:09:38       0.49   
2  2022-02-16T06:10:59  2022-02-16T18:09:41       0.50   
3  2022-02-17T06:10:48  2022-02-17T18:09:44       0.51   
4  2022-02-20T06:10:10  2022-02-20T18:09:48       0.58   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
1  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   
2  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   
3  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   
4  Rain, Partially cloudy  Partly cloudy throughout the day with a chance...   

   icon                                       stations  
0  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
1  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
2  rain  80222099999,80222399999,SKBO,80234099999,SKVV  
3  rain              80222099999,SKBO,80234099999,SKVV  
4  rain  80222099999,80222399999,SKBO,80234099999,SKVV

In [79]:
df_suba_rainy = pd.merge(df_suba_done, df_weather_p5, on=['name', 'datetime'], how='inner')

In [80]:
df_suba_rainy.head()

property_id    type_house business_type  bedrooms  bathrooms  parking_lots  \
0           10  departamento        Rental         1          2             1   
1           19  departamento        Rental         3          2             1   
2           19  departamento        Rental         3          2             1   
3           28  departamento        Rental         1          1             0   
4           28  departamento        Rental         1          1             0   

   services  balcony   pool  latitude  longitude localidad    city  \
0       NaN        0  False  4.723203 -74.053140      Suba  Bogotá   
1       NaN        0  False  4.767047 -74.048990      Suba  Bogotá   
2       NaN        0  False  4.767047 -74.048990      Suba  Bogotá   
3       3.0        0  False  4.753546 -74.093185      Suba  Bogotá   
4       3.0        0  False  4.753546 -74.093185      Suba  Bogotá   

                   region   country  scheduled_id                 begin_date  \
0  Región De Cundinamarca  Colombia        167764  2022-01-04T13:00:00-03:00   
1  Región De Cundinamarca  Colombia        176814  2022-01-14T15:00:00-03:00   
2  Región De Cundinamarca  Colombia        166495  2022-01-03T15:30:00-03:00   
3  Región De Cundinamarca  Colombia        178497  2022-01-17T15:30:00-03:00   
4  Región De Cundinamarca  Colombia        178497  2022-01-17T15:30:00-03:00   

                    end_date type_visit status                 name  \
0  2022-01-04T15:00:00-03:00      Visit   Done  4.7232027,-74.05314   
1  2022-01-14T17:00:00-03:00      Visit   Done  4.7670474,-74.04899   
2  2022-01-03T17:30:00-03:00      Visit   Done  4.7670474,-74.04899   
3  2022-01-17T18:00:00-03:00      Visit   Done  4.753546,-74.093185   
4  2022-01-17T18:00:00-03:00      Visit   Done  4.753546,-74.093185   

     datetime  tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike  \
0  2022-01-04     21.1      7.8  13.7          21.1           6.8       13.5   
1  2022-01-14     18.5      9.0  13.4          18.5           8.5       13.4   
2  2022-01-03     22.0      8.8  14.3          22.0           8.3       14.2   
3  2022-01-17     20.8      7.7  13.3          20.8           7.1       13.1   
4  2022-01-17     20.8      7.7  13.3          20.8           7.1       13.1   

    dew  humidity  precip  precipprob  precipcover preciptype  snow  \
0   8.8      74.6     0.7         100        29.17       rain   NaN   
1  10.3      82.8     3.9         100        45.83       rain   0.0   
2   7.5      70.1     1.7         100        54.17       rain   NaN   
3   9.6      80.9     2.4         100        41.67       rain   0.0   
4   9.6      80.9     2.4         100        41.67       rain   0.0   

   snowdepth  windgust  windspeed  winddir  sealevelpressure  cloudcover  \
0        NaN       NaN       25.6    166.4            1024.9        43.8   
1        0.0       6.1       17.9    224.5            1027.0        83.5   
2        NaN       NaN       22.3    178.6            1026.9        57.3   
3        0.0      10.4       22.3    187.9            1026.5        69.3   
4        0.0      10.4       22.3    187.9            1026.5        69.3   

   visibility  solarradiation  solarenergy  uvindex  severerisk  \
0        10.6           200.1         17.5        7         NaN   
1        10.5           132.9         12.0        5        10.0   
2        10.7           225.5         19.5        8         NaN   
3        10.2           204.3         17.7        7        30.0   
4        10.2           204.3         17.7        7        30.0   

               sunrise               sunset  moonphase  \
0  2022-01-04T06:05:34  2022-01-04T17:57:00       0.02   
1  2022-01-14T06:09:12  2022-01-14T18:01:32       0.45   
2  2022-01-03T06:05:13  2022-01-03T17:56:25       0.00   
3  2022-01-17T06:10:11  2022-01-17T18:02:58       0.50   
4  2022-01-17T06:10:11  2022-01-17T18:02:58       0.50   

               conditions                                        descr

## Respuesta

In [81]:
df_suba_rainy['temp'].mean()

14.371428571428577

La temperatura promedio para las visitas realizadas en la localidad de Suba es de 14,37 °C

In [82]:
df_suba_rainy['conditions'].unique()

array(['Rain, Partially cloudy'], dtype=object)